### PR TITLE
[DataGrid] Flex columns do not take into account "checkboxSelection" prop

### LINKED ISSUE #611: [DataGrid] Fluid column does not take into account checkbox
- [x] The issue is present in the latest release.
- [x] I have searched the [issues](https://github.com/mui-org/material-ui-x/issues) of this repository and believe that this is not a duplicate.

## Current Behavior 😯

The `XGrid` fluid column support using the `flex` prop does not take into account the checkbox width when `checkboxSelection` is true.  See [Codesandbox](https://codesandbox.io/s/material-ui-data-grid-bug-csdge?file=/src/App.tsx).  Uncomment the `checkboxSelection` prop and the grid renders the correct width, but with `checkboxSelection` the grid has a horizontal scroll equal to the checkbox width.

## Expected Behavior 🤔

Grid should have a width 100% of it's container with equal flex item widths even with `checkboxSelection` true.

## Steps to Reproduce 🕹

See above and Codesandbox to reproduce.

## Your Environment 🌎

| Tech          | Version |
| ------------- | ------- |
| Material-UI X | v4.0.0-alpha.10  |
| React         |    17     |
| Browser       |    Chrome     |
| TypeScript    |    Y     |

### PR DESCRIPTION
**Pull request by DanailH**
Fixes #611

### CODE REVIEW

File: packages/grid/data-grid/src/DataGrid.test.tsx [hunk 1]
```diff
@@ -297,6 +297,50 @@ describe('<DataGrid />', () => {
...
+        const availableColumnsWidth = checkboxSelectionColDef.width
+          ? columnsHeader!.offsetWidth - checkboxSelectionColDef!.width
+          : columnsHeader!.offsetWidth;
+        let columnsWidth = 0;
+
+        allColumnsExceptCheckbox.forEach((col) => {
+          columnsWidth += (col as HTMLElement).offsetWidth;
+        });
+
+        expect(availableColumnsWidth).to.equal(columnsWidth);
```
Comment (oliviertassinari) [inline packages/grid/data-grid/src/DataGrid.test.tsx:342]: -2 to account for the border as done in XGrid.

```suggestion
        expect(Array.from(document.querySelectorAll('[role="columnheader"]')).reduce((width, item) => width + item.clientWidth, 0)).to.equal(200 - 2);
```

Comment (DanailH) [reply packages/grid/data-grid/src/DataGrid.test.tsx:342]: Yeah I saw that but it was either hardcoding the 2 or do what I did and select the DOM elements 😅. I'm ok with this tho.

File: packages/storybook/src/stories/grid-columns.stories.tsx [hunk 2]
```diff
@@ -222,6 +222,23 @@ export function NewColumnTypes() {
   );
 }
 
+export const FlexColumnsWithCheckbox = () => {
```
Comment (oliviertassinari) [inline packages/storybook/src/stories/grid-columns.stories.tsx:225]: It feels like the duplication of test cases and stories aren't necessary. Maybe we could have the tests export a component and have the storybook display it (putting tests first)

**Review by oliviertassinari — APPROVED** [review]

File: packages/storybook/src/stories/grid-columns.stories.tsx (hunk 2 above)
Comment (DanailH) [reply packages/storybook/src/stories/grid-columns.stories.tsx:225]: I agree, it does look like duplication. I can remove the story, it is also a corner case, it doesn't really deserve its own story especially when it is covered by unit tests?

File: packages/grid/_modules_/grid/hooks/features/columns/useColumns.ts [hunk 3]
```diff
@@ -21,23 +21,27 @@ function mapColumns(
   columns: Columns,
   columnTypes: ColumnTypesRecord,
   viewportWidth: number,
+  withCheckboxSelection = false,
 ): Columns {
+  let availableViewportWidth = withCheckboxSelection
+    ? viewportWidth - checkboxSelectionColDef.width!
```
Comment (dtassone) [inline packages/grid/_modules_/grid/hooks/features/columns/useColumns.ts:27]: what if the user resizes the checkbox column?

File: packages/storybook/src/stories/grid-columns.stories.tsx (hunk 2 above)
Comment (oliviertassinari) [reply packages/storybook/src/stories/grid-columns.stories.tsx:225]: I'm not sure about removing the story, I was more concerned about the duplication of code.

File: packages/grid/_modules_/grid/hooks/features/columns/useColumns.ts [hunk 4]
```diff
@@ -57,7 +61,7 @@ function hydrateColumns(
   logger: Logger,
 ): Columns {
   logger.debug('Hydrating Columns with default definitions');
-  let mappedCols = mapColumns(columns, columnTypes, viewportWidth);
+  let mappedCols = mapColumns(columns, columnTypes, viewportWidth, withCheckboxSelection);
   if (withCheckboxSelection) {
     mappedCols = [checkboxSelectionColDef, ...mappedCols];
```
Comment (dtassone) [inline packages/grid/_modules_/grid/hooks/features/columns/useColumns.ts:66]: Why don't you swap this row with row 60 so when you do check the columns you already have the checkbox added?

File: packages/grid/_modules_/grid/hooks/features/columns/useColumns.ts (hunk 3 above)
Comment (DanailH) [reply packages/grid/_modules_/grid/hooks/features/columns/useColumns.ts:29]: Not a problem, resizing it behaves just like resizing any other column 😉

File: packages/grid/_modules_/grid/hooks/features/columns/useColumns.ts (hunk 4 above)
Comment (DanailH) [reply packages/grid/_modules_/grid/hooks/features/columns/useColumns.ts]: I removed the `mapColumns` function, it's not needed anymore. Now the logic is contained within the `hydrateColumns `.

**Review by dtassone — APPROVED** [review]
